internal/factsengine/gatherers/packageversion.go

```diff
@@ -120,13 +135,71 @@ func executeZypperVersionCmpCommand(
 func executeRpmVersionRetrieveCommand(
 	executor utils.CommandExecutor,
 	packageName string,
-) (string, *entities.FactGatheringError) {
-	rpmOutput, err := executor.Exec("/usr/bin/rpm", "-q", "--qf", "%{VERSION}", packageName)
+) ([]packageVersion, *entities.FactGatheringError) {
+	rpmOutputBytes, err := executor.Exec("/usr/bin/rpm", "-q", "--qf", packageVersionQueryFormat, packageName)
+
+	rpmOutput := string(rpmOutputBytes)
+
 	if err != nil {
-		gatheringError := PackageVersionRpmCommandError.Wrap(string(rpmOutput) + err.Error())
+		gatheringError := PackageVersionRpmCommandError.Wrap(rpmOutput + err.Error())
 		log.Error(gatheringError)
-		return "", gatheringError
+		return nil, gatheringError
+	}
+
+	installedVersions := []packageVersion{}
+
+	for _, detectedVersionLine := range strings.Split(rpmOutput, "\n---\n") {
+		if detectedVersionLine == "" {
+			continue
+		}
+
+		packageVersionInfo, err := envparse.Parse(strings.NewReader(detectedVersionLine))
+
+		if err != nil {
+			parsingError := fmt.Sprintf("Unable to parse rpm output: %s, output:%s", err.Error(), rpmOutput)
+			return nil, PackageVersionRpmCommandError.Wrap(parsingError)
+		}
+
+		packageVersionRequiredKeys := []string{installationTimeKey, packageVersionKey}
+
+		detectedPackageInstallationTime, installationTimeFound := packageVersionInfo[installationTimeKey]
+		detectedPackageVersion, packageVersionFound := packageVersionInfo[packageVersionKey]
+
+		if !installationTimeFound || !packageVersionFound {
```

Comment:
Yes, this is defensive :sweat_smile: I had the same thoughts.
It would mean that rpm output is messed up, and I am not sure if it would happen.

We could simplify by assuming those information would always be available and remove this if.

What do you prefer?

Reply:
If you don't do the checking of this, what will happen is:
- If the `VERSION` is missing, it will fill the version with an empty string
- If the `INSTALLEDTIME` is missing, the `ParseInt` fails.

Thinking that it is an scenario that is not possible (or let's say extremely unlikely unless rpm is broken) I would prefer to have a slimmer code. Specially on the test code

But well, as it is already implemented...
Your call, I'm fine with both solutions

Reply:
Makes sense to me. Adjusted to a slightly less defensive approach.